File: Jiho/week6/day27/택배 배달과 수거하기.js
function solution(cap, n, deliveries, pickups) {
  let answer = 0;

  let deliveryIndex = n - 1;
  let pickupIndex = n - 1;

  while (deliveryIndex >= 0 || pickupIndex >= 0) {
    while (deliveryIndex >= 0 && deliveries[deliveryIndex] === 0)
      deliveryIndex--;
    while (pickupIndex >= 0 && pickups[pickupIndex] === 0) pickupIndex--;

    let maxDistance = Math.max(deliveryIndex, pickupIndex) + 1;
    if (maxDistance === 0) break;

    answer += maxDistance * 2;

    let currentDelivery = 0;
    let currentPickup = 0;

    while (deliveryIndex >= 0 && currentDelivery < cap) {
      let amount = Math.min(cap - currentDelivery, deliveries[deliveryIndex]);
      deliveries[deliveryIndex] -= amount;
      currentDelivery += amount;
      if (deliveries[deliveryIndex] === 0) deliveryIndex--;
    }

    while (pickupIndex >= 0 && currentPickup < cap) {
      let amount = Math.min(cap - currentPickup, pickups[pickupIndex]);
      pickups[pickupIndex] -= amount;
      currentPickup += amount;
      if (pickups[pickupIndex] === 0) pickupIndex--;
    }
  }

  return answer;
}

File: Jin/week6/day27/택배배달과수거하기.java
package Jin.week6.day27;

public class 택배배달과수거하기 {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;

        int d = cap, p = 0;

        for (int i = n - 1; i >= 0; i--) {
            d -= deliveries[i];
            p += pickups[i];

            while (d < 0 || p > cap) {
                d += cap;
                p -= cap;
                answer += (i + 1) * 2;
            }
        }

        return answer;
    }
}
